crates/wast/src/core/binary.rs

```diff
@@ -625,6 +671,64 @@ impl Encode for Data<'_> {
     }
 }
 
+// Handle the case of branch hints here.
+// - Iterate over the instructions and get the hints
+// - Create the vector of branch hints and return it while encoding expressions
+impl EncodeFunction for Func<'_> {
+    fn encode_function(&self, e: &mut Vec<u8>) -> (&Func<'_>, Vec<BranchHintStruct>) {
+        assert!(self.exports.names.is_empty());
+        let mut tmp = Vec::new();
+        let (expr, locals) = match &self.kind {
+            FuncKind::Inline { expression, locals } => (expression, locals),
+            _ => panic!("should only have inline functions in emission"),
+        };
+
+        locals.encode(&mut tmp);
+        let tuple_vector = expr.encode_expr(&mut tmp);
+        tmp.len().encode(e);
+        e.extend_from_slice(&tmp);
+
+        // Use the tuple collection to build the branch hints collection
+        let mut branch_hints: Vec<BranchHintStruct> = Vec::new();
+        // Previous Instruction during iteration
+        let mut previous : Option<&Instruction> = None;
+        for (i, el) in tuple_vector.iter() {
+            match i {
+                Instruction::BranchHintAnnotation(bha) => {
+                    let bh = BranchHintStruct {
+                        branch_offset: (*el).try_into().unwrap(),
+                        reserved_byte: 1,
+                        branch_hint_value: bha.value.try_into().unwrap(),
+                    };
+
+                    branch_hints.push(bh);
+                }
+                _ => {
+                }
+            }
+            // Branch hints are only allowed before `br_if` and `if` instructions.
+            if previous.is_some() {
+                match previous {
+                    Some(Instruction::BranchHintAnnotation(_)) => {
+                        // If previous was a branch hint annotation, check that
+                        // the current instruction is indeed a `br_if` or a `if`
+                        match i {
+                            Instruction::BrIf(_) | Instruction::If(_) => {},
+                            _ => {
+                                panic!("Branch hint annotations must be followed by `br_if` or `if` instructions");
```

Comment:
Currently `wast` tries to avoid panicking wherever possible and instead defer errors like this to validation and/or compilation. Could this be silently ignored perhaps if it's attached to the wrong instruction?